libparsec/crates/platform_storage/src/web/certificates.rs

```diff
@@ -48,11 +167,114 @@ impl<'a> PlatformCertificatesStorageForUpdateGuard<'a> {
+                certificate_timestamp: timestamp.get_f64_with_us_precision(),
+                certificate: encrypted.into(),
+                certificate_type: certificate_type.into(),
+                filter1,
+                filter2,
+            },
+            &self.transaction,
+        )
+        .await
     }
 
     pub async fn get_last_timestamps(&mut self) -> anyhow::Result<PerTopicLastTimestamps> {
-        todo!()
+        let mut common_last_timestamp = None;
+        let mut sequester_last_timestamp = None;
+        let mut realm_certifs = Vec::new();
+        let mut shamir_recovery_last_timestamp = None;
+
+        for ty in COMMON_CERTIFICATES {
+            common_last_timestamp = Certificate::get_values(
+                &self.transaction,
+                CertificateFilter {
+                    certificate_type: ty.into(),
+                    filter1: None,
+                    filter2: None,
+                },
+            )
+            .await?
+            .into_iter()
+            .map(|certif| DateTime::from_f64_with_us_precision(certif.certificate_timestamp))
+            .chain(common_last_timestamp)
+            .max_by(|x, y| x.cmp(y));
+        }
+
+        for ty in SEQUESTER_CERTIFICATES {
+            sequester_last_timestamp = Certificate::get_values(
+                &self.transaction,
+                CertificateFilter {
+                    certificate_type: ty.into(),
+                    filter1: None,
+                    filter2: None,
+                },
+            )
+            .await?
+            .into_iter()
+            .map(|certif| DateTime::from_f64_with_us_precision(certif.certificate_timestamp))
+            .chain(sequester_last_timestamp)
+            .max_by(|x, y| x.cmp(y));
+        }
+
+        for ty in REALM_CERTIFICATES {
+            realm_certifs.extend(
+                Certificate::get_values(
+                    &self.transaction,
+                    CertificateFilter {
+                        certificate_type: ty.into(),
+                        filter1: None,
+                        filter2: None,
+                    },
+                )
+                .await?
+                .into_iter()
+                .map(|certif| {
+                    (
+                        certif.filter1,
+                        DateTime::from_f64_with_us_precision(certif.certificate_timestamp),
+                    )
+                }),
+            );
+        }
+
+        realm_certifs.sort_by(|x, y| x.1.cmp(&y.1));
+
+        let per_realm_last_timestamps = realm_certifs
+            .into_iter()
+            .map(|x| {
+                x.0.ok_or(anyhow::anyhow!("Missing realm id"))
+                    .and_then(|id| {
+                        VlobID::from_hex(&id)
+                            .map(|id| (id, x.1))
+                            .map_err(|e| anyhow::anyhow!(e))
+                    })
+            })
+            .collect::<Result<HashMap<_, _>, _>>()?;
```

Comment:
Why do you need to sort the `realm_certifs` if you put them in a `HashMap` ?

Reply:
Because It need the last timestamp for each realm

Reply:
Why do you need the most recent entry to be inserted last ?

From what I know `HashMap` don't provide any guaranty about the iteration order, what I'm missing ?

Reply:
The order is guaranted by the sort as you mention

Reply:
Now I understand

> the last inserted is the most recent

You need an `Hasmap<VlobID, DateTime>` where when you have the same `VlobID` you only want to keep the most recent date.